lib/Sema/CSSimplify.cpp

```diff
@@ -12972,16 +12972,11 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyFixConstraint(
     getDefaultDecompositionOptions(flags) | TMF_ApplyingFix;
   switch (fix->getKind()) {
   case FixKind::ForceOptional: {
-    SmallVector<Type, 4> unwraps1;
-    type1->lookThroughAllOptionalTypes(unwraps1);
-
-    SmallVector<Type, 4> unwraps2;
-    type2->lookThroughAllOptionalTypes(unwraps2);
-
-    auto impact = unwraps1.size() != unwraps2.size()
```

Comment:
Still have to think about this a bit more, in one specific case where fromObjectType is a type variable we still record force optional fix with from and to with same level of optional if object types match which means toObjectType is also a type variable probably. Have to look a bit more

Reply:
Can we "delay" a fix here by returning "unsolved" until the type variable that represents object type is resolved?

Reply:
This makes sense, I'll have more time over the weekend to work on that. Let's try that and see :)

Reply:
I just looks like we somehow already "delaying" but only for `Equal` constraints https://github.com/apple/swift/blob/99dcabd1a62475059dea61f582ab180f80db67b0/lib/Sema/CSSimplify.cpp#L4456
not in the sense that we return "unsolved" and put the constraint back to wait until a fixed type is bound, but in that force unwrap is not recorded if fixed types aren't bound for equality constraints only. To me is still a bit confusing why we allowed equal number of unwraps record a force optional fix when type variables are involved because it could or not be an optional so force optional does not look correct in this case. But is possible I'm missing something.

Reply:
Yeah, that doesn't make much sense to report a fix when there is no unwrapping needed seems like that should be an assert instead.

Reply:
We need to figure out in what situations that happens (if any) and address them at the source.

Reply:
> Yeah, that doesn't make much sense to report a fix when there is no unwrapping needed seems like that should be an assert instead.

That is actually how this was caught because `ForceOptional` now has an assert on optional count diff so that is why I removed the type variable check from https://github.com/apple/swift/blob/3c4dd9fad48a0843b43a444a77e18bc65d9e8e54/lib/Sema/CSSimplify.cpp#L4571

So that is why I was asking about why we only exit early for Equal constraints above, but allowing typevars otherwise which potentially ends up in a equal optionality recorded as ForceOptional. 

> We need to figure out in what situations that happens (if any) and address them at the source.

Sure, I'll try go back to this, was a bit confusing. Thanks for the input =]

Reply:
> So that is why I was asking about why we only exit early for Equal constraints above

I think the reason there is partially to optional evaluation expressions which preserve optionality, try extending it and see what fails.

Reply:
Sorry for taking a while to respond 😅 

>  try extending it and see what fails.

Removing the `Equal` match kind from the check only breaks on one case in the whole test suite involving a dependent member with optional base https://godbolt.org/z/jv48h8W63. But in this case makes sense to record the force optional fix because we have a `$T0?` bind `Int`.

Reply:
No worries! I guess we could narrow the check to allow it down to that situation only then?

Reply:
I'm thinking a bit about this, this condition seems fine we for sure could narrow it but that makes sense the way it is.
The problematic case I think is this `!fromObjectType->is<TypeVariableType>()`
https://github.com/apple/swift/blob/6676fbfd19781dd6ed1d59bb84146aff336c24d4/lib/Sema/CSSimplify.cpp#L4571
I don't think we should allow a force optional fix when fromObjType is a type variable with less or equal unwraps then to. Because we can't know for sure it would be an optional(as previous comment said "it may be" but it cannot). I already removed that and it seems not to break any diagnostic in the suit and neither the new assert in ForceOptional ` assert(fromOptionals > toOptionals)` so `getUnwrapCount()` will always be > 0. That makes me think that it should be ok. Do you think that makes sense? 
I'll run the full suit of tests in CI just to make sure.

Reply:
> I don't think we should allow a force optional fix when fromObjType is a type variable with less or equal unwraps then to. 

Yes, that is correct. The check you mentioned is checking exactly that, it only allows the fix if "from" is not a type variable, that the situation when we know for sure that there is a missing unwrap.

Reply:
You mean this one https://github.com/apple/swift/blob/6676fbfd19781dd6ed1d59bb84146aff336c24d4/lib/Sema/CSSimplify.cpp#L4571
? If I understand correctly this is the other way around, like only exit early in case from less or equal and from is not a type var which is what is allowing type vars with equal number of unwraps being recorded as force optional. Let me take a look again to make sure, is possible I miss understood the condition

Reply:
Looks like it's indeed the other way around, and that is what was allowing type vars to ignore this condition `fromUnwraps <= toUnwraps`.

Reply:
You are right, it should be the other way around - the comment states one thing but the code does the other :)

Reply:
> No worries! I guess we could narrow the check to allow it down to that situation only then?

Just did that adjustment, just don't record force optional if type variables are involved expect for that case. Let me know what you think.